### ISSUE TITLE
PCC drop observed in Conv2d + BatchNorm2d ops: PCC = 0.9880387783050537

### ISSUE THREAD

**Issue by kamalrajkannan78** [tenstorrent/tt-metal#34229]
### Describe the bug

- PCC drop observed in Conv2d + BatchNorm2d ops: PCC = 0.9880387783050537 
- The issue cannot be further isolated, as both ops achieve PCC > 0.99 when run independently:
  - Conv2d - 0.9997682571411133
  - BatchNorm2d - 0.999877393245697
- Openpose V2 Model affected by this drop - https://github.com/tenstorrent/tt-xla/issues/2417

### Steps to reproduce the issue

```
git checkout kkannan/dec15_openpose_conv2d_bn2d_pcc_drop_metal
git submodule update --init --recursive
git lfs pull --include '*.pt tensors/*'

pip install pytorchcv==0.0.73

# Conv2d + BatchNorm2d
pytest tests/ttnn/unit_tests/operations/conv2d_bn2d/test_sanity.py -svv

# conv2d 
pytest tests/ttnn/unit_tests/operations/conv2d_bn2d/test_conv2d.py -svv 

# Batchnorm2d
pytest tests/ttnn/unit_tests/operations/conv2d_bn2d/test_bn2d.py -svv 
```

### Logs 

- [dec15_conv2d_bn2d_pcc_drop_f.log](https://github.com/user-attachments/files/24165149/dec15_conv2d_bn2d_pcc_drop_f.log)
- [dec15_conv2d_pcc.log](https://github.com/user-attachments/files/24168961/dec15_z_conv2d_pcc.log)
- [dec15_bn2d_pcc.log](https://github.com/user-attachments/files/24168960/dec15_zz_bn2d_pcc.log)

### Note 

- To confirm whether the PCC drop originates from the Conv2d+BatchNorm2d operations themselves or from surrounding ttnn operations (reshape, permute), a sanity test (tests/ttnn/unit_tests/operations/conv2d_bn2d/test_sanity.py ) was created where:
  - Only Conv2d and BatchNorm2d ops execute on ttnn device
  - All preprocessing/postprocessing operations (reshape, permute) execute on CPU using torch
- To get the results from the original TTNN repo got from [emitpy tool](https://docs.tenstorrent.com/tt-xla/emitpy_tutorial.html) run `pytest tests/ttnn/unit_tests/operations/conv2d_bn2d/test_conv2d_bn2d.py -svv` - [dec15_conv2d_bn2d_pcc_drop_org.log](https://github.com/user-attachments/files/24165150/dec15_conv2d_bn2d_pcc_drop_org.log) 
- Note : Drop is same in both cases.
- only the problematic ops are executed on ttnn device in op tests too.

**Comment by bbradelTT** [comment]
@kamalrajkannan78 can you please try enabling fp32 dest acc for batch norm, and possibly conv?

If you don't, then everything will go through tf32 and accuracy is going to be bad for many inputs. The call to conv beforehand probably skews inputs in a way that's leading to inaccuracy.

**Comment by kamalrajkannan78** [comment]
> [@kamalrajkannan78](https://github.com/kamalrajkannan78) can you please try enabling fp32 dest acc for batch norm, and possibly conv?
> 
> If you don't, then everything will go through tf32 and accuracy is going to be bad for many inputs. The call to conv beforehand probably skews inputs in a way that's leading to inaccuracy.

Thanks for the suggestion, @bbradelTT.

- I tried enabling fp32 destination accumulation for the following cases, and below are the corresponding PCC values:
  - ttnn.conv2d (fp32 dest):` 0.7963908910751343` -[dec16_conv2d_fp32_dest.log](https://github.com/user-attachments/files/24186227/dec16_z_conv2d_fp32_dest.log) 
  - ttnn.batch_norm (fp32 dest):` 0.9880310893058777` - [dec16_bn2d_fp32_dest.log](https://github.com/user-attachments/files/24186225/dec16_z_bn2d_fp32_dest.log)
  - ttnn.conv2d (fp32 dest) + ttnn.batch_norm (fp32 dest): `0.796806812286377` - [dec16_conv2d_bn2d_fp32_dest.log](https://github.com/user-attachments/files/24186226/dec16_z_conv2d_bn2d_fp32_dest.log)

- Unfortunately, enabling fp32 dest did not improve the PCC in these scenarios.  Note : the input dtype is bfloat16.
- Also, I should not add any additional configuration to the repro code, since it is TTNN code generated directly from the model’s TTNN graph. Adding extra configuration would not reflect the original problematic scenario.

### Repro 

```
git checkout kkannan/dec15_openpose_conv2d_bn2d_pcc_drop_metal
git submodule update --init --recursive
git lfs pull --include '*.pt tensors/*'

pip install pytorchcv==0.0.73

# ttnn.conv2d (fp32 dest)
pytest tests/ttnn/unit_tests/operations/conv2d_bn2d/test_sanity_conv2d_fp32.py

# ttnn.batch_norm (fp32 dest)
pytest tests/ttnn/unit_tests/operations/conv2d_bn2d/test_sanity_bn2d_fp32.py -svv 

# ttnn.conv2d (fp32 dest) + ttnn.batch_norm (fp32 dest)
pytest tests/ttnn/unit_tests/operations/conv2d_bn2d/test_sanity_conv2d_bn2d_fp32.py -svv

```

**Comment by kamalrajkannan78** [comment]
### Root Cause Analysis

- BatchNorm amplifies the PCC drop, reducing it from `0.9997682571411133` (Conv2d output) to `0.9880388975143433 `(after BatchNorm.)
-  To validate this, I replaced ttnn.batch_norm with the Torch implementation. In this setup, Conv2d is the only TTNN op, yet the final PCC is still `0.9880388975143433`.  [dec16_conv2d_tt_bn2d_cpu.log](https://github.com/user-attachments/files/24186598/dec16_z_conv2d_tt_bn2d_cpu.log)
- To further confirm, I compared the tensors before BatchNorm, where the PCC remains `0.9997682571411133`. [dec16_drop_amplification.log](https://github.com/user-attachments/files/24186602/dec16_z_drop_amplification.log)
- Even when applying CPU BatchNorm on the ttnn.conv2d output, the PCC drop persists. This suggests the issue originates from the Conv2d output accuracy, which is then magnified by BatchNorm.
- Based on this, it appears that improving Conv2d accuracy could potentially resolve the issue. However, I’m not sure whether further accuracy improvements are feasible at this point as it already >0.99
-  Since this appears to be a unique scenario, I’d appreciate your thoughts or guidance on how best to proceed. @pavlejosipovic @bbradelTT 

### Repro

```
git checkout kkannan/dec15_openpose_conv2d_bn2d_pcc_drop_metal
git submodule update --init --recursive
git lfs pull --include '*.pt tensors/*'

pip install pytorchcv==0.0.73

# validation 
tests/ttnn/unit_tests/operations/conv2d_bn2d/test_conv2d_tt_bn2d_cpu.py
tests/ttnn/unit_tests/operations/conv2d_bn2d/test_amplification.py
```

**Comment by bbradelTT** [comment]
@kamalrajkannan78 can you please try with:
```
    compute_config = ttnn.init_device_compute_kernel_config(
        device.arch(),
        math_fidelity=ttnn.MathFidelity.HiFi2,
        math_approx_mode=False,
        fp32_dest_acc_en=True,  # ?~F~P Enable FP32 dest
    )
```
I'm getting
```
Output PCC: 0.9993454813957214
```

**Comment by kamalrajkannan78** [comment]
> [@kamalrajkannan78](https://github.com/kamalrajkannan78) can you please try with:
> 
> ```
>     compute_config = ttnn.init_device_compute_kernel_config(
>         device.arch(),
>         math_fidelity=ttnn.MathFidelity.HiFi2,
>         math_approx_mode=False,
>         fp32_dest_acc_en=True,  # ?~F~P Enable FP32 dest
>     )
> ```
> 
> I'm getting
> 
> ```
> Output PCC: 0.9993454813957214
> ```

@bbradelTT 

After applying the above configuration, I tested the following cases and observed the corresponding PCC values:

- ttnn.conv2d (fp32 dest): 0.9993295669555664 - [dec17_conv2d_fp32_dest.log](https://github.com/user-attachments/files/24204351/dec17_conv2d_fp32_dest.log)
- ttnn.batch_norm (fp32 dest):  0.9880310893058777 - [dec17_bn2d_fp32_dest.log](https://github.com/user-attachments/files/24204350/dec17_bn2d_fp32_dest.log)
- ttnn.conv2d (fp32 dest) + ttnn.batch_norm (fp32 dest): 0.9993489980697632 - [dec17_conv2d_bn2d_fp32_dest.log](https://github.com/user-attachments/files/24204352/dec17_conv2d_bn2d_fp32_dest.log)

Case 1 & 3 have pcc>0.99 now. I can add the mentioned config to make the test pass but as I mentioned [here](https://github.com/tenstorrent/tt-metal/issues/34229#issuecomment-3659339361), I should not add any additional configuration to the repro code, since it is TTNN code generated directly from the model’s TTNN graph. 
Is there a possibility of making this configuration as default in future?

**Comment by nvukobratTT** [comment]
We're looking into making some of them default on the compiler level, so I hope that these types of issues are mostly going to be resolved for the basic runs.

cc @dgolubovicTT

**Comment by bbradelTT** [comment]
Once the pipelines are ok, I'll try to create a PR to enable fp32 dst acc by default.

**Comment by bbradelTT** [comment]
@pavlejosipovic increasing batch_norm default accuracy is not sufficient to get the full benefit. The default conv2d kernel config would need to be changed.

```
diff --git a/ttnn/cpp/ttnn/operations/conv/conv2d/conv2d_utils.cpp b/ttnn/cpp/ttnn/operations/conv/conv2d/conv2d_utils.cpp
index 23cc17f440..91d34d096d 100644
--- a/ttnn/cpp/ttnn/operations/conv/conv2d/conv2d_utils.cpp
+++ b/ttnn/cpp/ttnn/operations/conv/conv2d/conv2d_utils.cpp
@@ -531,7 +531,7 @@ SkipMcast conv_skip_mcast(const Conv2dParallelizationConfig& parallelization_con
 }
 
 DeviceComputeKernelConfig get_conv_default_compute_kernel_config(MeshDevice* device) {
-    return init_device_compute_kernel_config(device->arch(), std::nullopt, MathFidelity::HiFi4, true, false, false);
+    return init_device_compute_kernel_config(device->arch(), std::nullopt, MathFidelity::HiFi4, false, true, true);
 }
```

What are your plans for accurate by default convs? Would you be okay with me making the above change?

**Comment by s-jovic** [comment]
Hey @bbradelTT, we don't have any plans for improving accuracy by default. @astancovTT recently fixed accuracy issues with fp32 inputs by setting fp32 dest acc true in this case (https://github.com/tenstorrent/tt-metal/commit/2f8845059f8e6b83f13804a0d6eb559cc38c9936).

Changing the default compute kernel config would probably result in perf regressions in many models. @kamalrajkannan78 can you elaborate on why it's not possible to add any additional configuration?

**Comment by kamalrajkannan78** [comment]
> @kamalrajkannan78 can you elaborate on why it's not possible to add any additional configuration?

- The Openpose model shows a PCC of `PCC=0.9080588817596436` in tt-xla - https://github.com/tenstorrent/tt-xla/issues/2417
- The PCC drop originates from the [Conv2d + BatchNorm2d](https://github.com/osmr/pytorchcv/blob/7f31dc8fe521e99012ac863a946e10a314118c16/pytorchcv/models/common/conv.py#L281C1-L283C27) - `PCC=0.9880387783050537`
- The attached TTNN sanity repro is part of the actual model graph. While it is possible to add additional configuration to make this sanity test pass, the full model runs with the default configuration, where the PCC drop is still observed.

**Comment by nvukobratTT** [comment]
@kamalrajkannan78 @s-jovic we're now introducing the ability to update FP32 accumulation and MF for operations that don't have FP32 as well. This might be a solution for this specific use case. 

@kamalrajkannan78, once those changes land on the compiler, let's test them out and share the findings here. Feel free to follow these PRs:
- https://github.com/tenstorrent/tt-xla/pull/2844/changes

**Comment by kamalrajkannan78** [comment]
@nvukobratTT @bbradelTT @s-jovic 

- Changes from this [PR](https://github.com/tenstorrent/tt-xla/pull/2844) landed on `tt-xla`.
-  `Conv2d + Batchnorm2d` sanity passing with `PCC=0.9994961321298823` & Model's PCC slightly improved from `0.9080588817596436` to `0.9180459677147291`. 
- For complete context, pls checkout https://github.com/tenstorrent/tt-xla/issues/2417#issuecomment-3771108482
- This Indicates that Model need to be bisected again to find the problematic op
- Since the previously identified problematic ops are now passing with valid PCC on tt-xla, I am closing this ticket.
- Thanks to everyone for the help and suggestions.
